### PR TITLE
Simplify KBPKB endgame with opposite bishops

### PR DESCRIPTION
**Pull request by snicolet**
When we reach a position with only two opposite colored bishops and one pawn on the board, current master would give it a scale factor of 9/64=0.14 in about one position out of 7200, and a scale factor of 0.0 in the 7199 others. The patch gives a scale factor of 0.0 in 100% of the cases.

STC:
LLR: 2.96 (-2.94,2.94) [-3.00,1.00]
Total: 55845 W: 11467 L: 11410 D: 32968
http://tests.stockfishchess.org/tests/view/5abc585f0ebc5902926cf15e

LTC: 
LLR: 2.95 (-2.94,2.94) [-3.00,1.00]
Total: 11915 W: 1852 L: 1719 D: 8344
http://tests.stockfishchess.org/tests/view/5abc7f750ebc5902926cf18c

This patch is a side effect of an ongoing effort to improve the SF evaluation of positions involving a pair of opposite bishops. See the GitHub diff of this LTC test which almost passed at sprt[0..5] for a discussion: http://tests.stockfishchess.org/tests/view/5ab9030b0ebc5902932cbf93

No functional change (at small bench depths)

### CODE REVIEW

**Comment by Rocky640** [comment]
@snicolet` in about one position out of 7200`

I don't know about how good the removed code is or not. (I did not even looked at it).

But which set of positions did you take ?
I would be very careful. If you get this from a long bench, all your 7200 positions may come more or less from the same starting position or two...

**Comment by Hanamuke** [comment]
Either way it passed LTC easily.

**Comment by Rocky640** [comment]
OK, I did some homework.

I created a auto generator which produced 100000 positions for this KBPkb (and opposed bishop and pawn on rank 5 and above), black to move, not in check, and no white piece "en prise"

This ending is very drawish. My stopping criteria was once a score <=0.5 is produced in case of a known draw or a score >=2 in case of known win).

According to this criteria, it can be "solved" with less than 10000 nodes on 99.961% of the randomly generated positions. I generated only positions with a pawn on rank 5 and above,

Here I show some statistics for numbers of nodes needed to "solve" each positions.

I will try to produce, tomorrow, the same data using  this PR as my sf exe, and also some comparative data for a more complex ending such as same KBPkb with same color bishop

image

**Comment by Stefano80** [comment]
Stephane, what is your opinion about the endgame code? For years the guideline by Marco was: either remove all of them in one go or leave them in peace.

If this is changed I will reopen my proposal about rook and minors.

**Comment by snicolet** [comment]
@Rocky640 @Stefano80 

I have the impression that keeping the fake scale factor 9 in positions where the attacking side has (by definition) a passed pawn on ranks 5-7 (which are evaluated quite high by default), does in fact add **noise** in the evaluation which prevents us from making some cuts in the alpha-beta search, more noise than the very few false positive draws that I introduce by using scale factor 0.

Would you accept the patch if it passes a SPRT[0..4] at LTC?

**Comment by Stefano80** [comment]
@snicolet you understood me wrong. I am definitely for merging this AND any removal of useless endgame knowledge. I was only asking if you would reconsider my old about rooks and minors one after merging this.

I am anyway 100% for merging this with our without reconsideration of my old one.

**Comment by Stefano80** [comment]
This one is the old one, if you're interested https://github.com/official-stockfish/Stockfish/pull/1280

What I measured back then is that none of KmPKm helps, even if you put the together.

**Comment by Rocky640** [comment]
I changed my methodology. It makes more sense to use go depth x instead of go nodes x

Test  conditions
100000 were auto generated
all with weak side (=black) to play, and  no legal capture on next move, and no king in check in the start position
13 positions were a mate in one by black   (white king is cornered on a8 by own pawn)
Those 13 positions were added to the  "win" column
 
A position is considered "solved at depth  x" 
-if it is a TB draw and score reported for a new search at depth   x  is less than 0.5
-or if it is a TB win and score reported   is >2

According to this definition, and the data below, I consider this PR is identical to master to "solve for draw" and better than master to solve earlier for "wins"

image

Stay tuned for comparative data for a more difficult ending, KBPkb with same color bishops.

**Comment by jhellis3** [comment]
Just wanted to offer a comment on when this type of simplification makes sense (it does here)...

If the far most likely outcome is a draw, and the scaling function is not absolutely perfect, then it makes a great deal of sense to simplify to draw and allow the search to, hopefully, find a more likely (easier to convert) path to victory.

On the other hand, simplifying away conditions which return draw is pretty dangerous, as it allows search to go into forced draws thinking it is winning, possibly preferring those lines over other actually winning paths. 

TLDR: Ok to be overly pessimistic, not ok to be overly optimistic.

**Comment by jhellis3** [comment]
I would also add that it should then be possible to actually improve the strength of the code by identifying the rare situations which are won/lost with 100% certainty. However, as these situations will be very rare anyway, it would likely be impossible to measure an Elo gain with standard testing practices.

**Comment by Rocky640** [comment]
For the above PR' all the positions are "solved" before depth 10.

Here's my data for current master for the same color bishops.
This is not related to this PR, but shows that we are close to super accuracy in the opposed case.
Positions with pawns on rank 4 were not generated. 
image
